### PR TITLE
Feat/wizard

### PR DESCRIPTION
**Pull request by Si3rr4wow**
Ticket: #1 

https://user-images.githubusercontent.com/35689158/203349913-691d9601-ba04-4c70-9ee8-8c4b16008e9b.mov

### CODE REVIEW

File: src/wizard.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  initial: `openapi.${documentSyntax}`,
+  validate: (input) =>
+    !!input.match(
+      new RegExp(
+        `(.*)(?<!\\.|${documentSyntax === 'json' ? 'yaml' : 'json'})$`,
+      ),
+    ) ||
+    `Entry point extension must match chosen document syntax (you chose ${documentSyntax})`,
+  format: (input): string => {
+    if (input.match(/(\.json|\.yaml)$/)) {
```
Comment (robmorgan-tab) [inline src/wizard.ts:46]: Consider case sensitivity here, some people put their extensions in uppercase.

File: src/wizard.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+export const getEntryPointPrompt = (documentSyntax: string): PromptObject => ({
+  name: 'entryPoint',
+  type: 'text',
+  message: 'Entry point',
+  initial: `openapi.${documentSyntax}`,
+  validate: (input) =>
+    !!input.match(
+      new RegExp(
+        `(.*)(?<!\\.|${documentSyntax === 'json' ? 'yaml' : 'json'})$`,
```
Comment (robmorgan-tab) [inline src/wizard.ts:41]: If you're using negative lookbehinds in regular expressions, make sure you include an `engines` field in the `package.json`, as older versions of node don't support this syntax.

File: src/wizard.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    !!input.match(/^[\w\-_]+$/) ||
+    'Project name may only contain letters, hyphens and underscores',
+})
+
+export const getDocumentSyntaxPrompt = (): PromptObject => ({
+  name: 'documentSyntax',
+  type: 'select',
+  message: 'Document syntax',
+  choices: [
+    { title: 'json', value: 'json' },
```
Comment (robmorgan-tab) [inline src/wizard.ts:27]: Just an idea to use the proper capitalisation in the human-readable labels.
```suggestion
    { title: 'JSON', value: 'json' },
    { title: 'YAML', value: 'yaml' },
```

File: src/wizard.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  } as ProjectOptions
+}
+
+const makeBoilerplate = (
+  processDirectory: string,
+  { entryPoint, projectName }: ProjectOptions,
+): void => {
+  let writeDirectory = '.'
+  if (processDirectory !== projectName) {
+    fs.mkdirSync(projectName)
```
Comment (robmorgan-tab) [inline src/wizard.ts:73]: Do you need to sanitise the project name at all?

File: src/wizard.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    writeDirectory += `/${projectName}`
+  }
+  fs.writeFileSync(`${writeDirectory}/${entryPoint}`, 'Hello Spec!')
+}
+
+export const init = async (): Promise<void> => {
+  const processDirectory = process.env.INIT_CWD?.split('/').reverse()[0] || ''
```
Comment (robmorgan-tab) [inline src/wizard.ts:80]: Something to consider is that _I think_ `INIT_CWD` isn't available by default on Windows systems, which is why it's an optional type.

File: src/wizard.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      const randomString = Math.random().toString(36).substring(2)
+      const projectNamePrompt = getProjectNamePrompt(randomString)
+      expect(projectNamePrompt.initial).toBe(randomString)
+    })
+  })
+
+  describe('validation', () => {
+    it('permits alphanumeric names', () => {
+      const projectNamePrompt = getProjectNamePrompt('.')
+      const randomString = Math.random().toString(36).substring(2)
```
[the remark this answers is not in the document]
Comment (Si3rr4wow) [reply src/wizard.test.ts:15]: if only `toString` could handle base 62 😪

File: src/wizard.ts (hunk 4 above)
Comment (Si3rr4wow) [reply src/wizard.ts:73]: the validation on the prompt is `/^[\w\-_]+$/`, think that suffices?

Comment (robmorgan-tab) [reply src/wizard.ts:73]: I'd suggest being explicit and confine the range to ASCII characters using: `/^[a-z0-9\-_]+$/i`

File: src/wizard.ts (hunk 5 above)
Comment (Si3rr4wow) [reply src/wizard.ts]: This is now the mech neater ` const processDirectory = path.parse(process.env.INIT_CWD || '').base
`

File: src/wizard.ts (hunk 4 above)
Comment (Si3rr4wow) [reply src/wizard.ts:74]: 👍  committed this suggestion

File: src/wizard.ts (hunk 2 above)
Comment (Si3rr4wow) [reply src/wizard.ts:42]: added >18.12.1, inline with the .nvmrc

File: src/wizard.ts (hunk 1 above)
Comment (Si3rr4wow) [reply src/wizard.ts]: added an `i`

File: src/wizard.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Si3rr4wow) [reply src/wizard.test.ts]: yup that works, added to test utils with the random string generator chain

**Review by robmorgan-tab — APPROVED** [review]

**Pull request merged**